frontend/public/components/utils/storage-class-dropdown.tsx

```diff
@@ -0,0 +1,201 @@
+        <span className="text-muted">{this.props.placeholder}</span>
+      ) : (
+        <LoadingInline />
+      );
+      this.setState({ title });
+    }
+
+    componentWillReceiveProps(nextProps) {
+      const { loaded, loadError, resources, selectedKey } = nextProps;
+
+      if (loadError) {
+        this.setState({
+          title: (
+            <div className="cos-error-title">Error Loading {nextProps.desc}</div>
+          ),
+        });
+        return;
+      }
+      if (!loaded) {
+        return;
+      }
+
+      const state = {
+        items: {},
+        title: {},
+      };
+
+      const noStorageClass = 'No storage class';
+      state.items[noStorageClass] = { kindLabel: '', name: noStorageClass };
+      _.each(resources, ({ data }, kindLabel) => {
+        _.reduce(
+          data,
+          (acc, resource) => {
+            acc[resource.metadata.name] = {
+              kindLabel,
+              name: resource.metadata.name,
+              description: resource.metadata.annotations ? resource.metadata.annotations.description : '',
+              default: resource.metadata.annotations ? resource.metadata.annotations['storageclass.beta.kubernetes.io/is-default-class'] === 'true' ? this.setState({defaultClass:resource.metadata.name}, ()=> {
+                return true;
+              }) : false : '',
+              accessMode: resource.metadata.annotations ? resource.metadata.annotations['storage.alpha.openshift.io/access-mode'] : '',
+              provisioner: resource.provisioner,
+              type: resource.parameters ? resource.parameters.type : '',
+              zone: resource.parameters ? resource.parameters.zone : '',
+            };
+            return acc;
+          },
+          state.items
+        );
+      });
+
+      const { defaultClass } = this.state;
+      if (defaultClass) {
+        // No Storage Class is no longer an option if there is a default
+        delete state.items[noStorageClass];
+      }
+
+      if (!this.props.loaded || !selectedKey || selectedKey === noStorageClass) {
+        state.title = <span className="text-muted">{nextProps.placeholder}</span>;
+      }
+
+      if (selectedKey && selectedKey !== noStorageClass) {
+        const item = state.items[selectedKey];
+        if (item) {
+          state.title = this.getTitle(item.kindLabel, item.name);
+        }
+      }
+
+      if (!selectedKey && defaultClass) {
+        const item = state.items[defaultClass];
+        if (item) {
+          state.title = this.getTitle(item.kindLabel, item.name);
+        }
+      }
+
+      this.setState(state);
+    }
+
+    shouldComponentUpdate(nextProps, nextState) {
+      if (_.isEqual(this.state, nextState)) {
+        return false;
+      }
+      return true;
+    }
+
+    autocompleteFilter = (text, item) => fuzzy(text, item.props.name);
+
+    getTitle = (kindLabel, name) => {
+      return kindLabel ? <ResourceName kind={kindLabel} name={name} /> : <span >{name}</span>;
```

Comment:
We should always know the kind label, correct? It's always StorageClass

Reply:
I don't have the kindLabel set on the "No Storage Class" option so that I do not show an icon. That is why I check it.  I can hardcode the ResourceName kind values though.